testing.md

```diff
@@ -0,0 +1,72 @@
+# Testing
+
+## Why write tests?
+We write automated tests to allow us to demonstrate that our software meets business requirements, and that it continues to do so as changes are made to the codebase.
+
+## When do I write tests?
+Almost all of our code should be tested to some degree, the question is less about "when" and more "to what level do I test this code?". There are a few points to consider to help with this decision making.
+
+### Net costs
+It's worth considering the net cost of tests before you write them. Simply put, ask yourself "Is this test likely to pay for itself over the course of it's lifetime?".
+
+If not, it could be that you're incurring a net cost to the business with a over-thorough test that needs to be **1.** understood, **2.** run as part of your CI and **3.** maintained. In this scenario it might be worth opting for a something less expensive that still gives the team an acceptable level of confidence that the code works as expected.
+
+### Keep the system in mind
+When making decisions about the scope of your testing, keep the system in mind that you're working on. For example:
+ - A single failure on a data-sensitive area of an API is potentially going to affect hundreds of customers and could expose customer data. This needs extensive, thorough, testing to ensure customer data is never exposed.
+ - Conversely, s single failure in a simple React component is only going to affect a single session at once, and might just incur a slight visual glitch or formatting issue. We probably want to write a quick test to check this works as expected, but testing every possible edge case is likely to cost us more than the test is worth.
+ - Library code should always be tested extensively It's designed to be reused many times across many systems in many use cases - we need to be confident this works as advertised no matter how it's used.
+
+## Types of Testing
+There are three major categories of testing that might be appropriate to use in your development.
+
+### Unit Testing
+[Unit testing](https://en.wikipedia.org/wiki/Unit_testing) involves testing code at a very granular level, usually testing a single function and everything that goes on within it, stubbing calls to other functions or modules to ensure we can accurately check the functionality of the unit. Taking making a cake as an analogy, you can think of unit testing as checking the quality of the individual ingredients.
+
+Unit tests are particularly useful for things like:
+ - Proving the functionality of complex algorithmic logic.
+ - Alerting developers to potential regressions during refactors in a "yes, I really did mean to do that there" kind of fashion.
```

Comment:
Unit tests only offer value for refactor if the refactor is within the unit. If for example you are refactoring a module and you've tested at the function level you will likely need to either throw out your tests or substantial rewrite them.